### DISCUSSION TITLE
How i can use my own cmake to build a code without using open_spiel as a shared libary ?

### DISCUSSION THREAD

**Discussion by LucasCelestinoSE** [deepmind/open_spiel#1184] [Q&A] [answered]
I'm trying to do cmake to build my own project.
I'm using EMSCRIPTEN and it does not have support for a shared libary, just the libaries inside the box...
I was left with only one alternative, using cmake. But I don't know where to start, I don't have that much experience using cmake, and there are several existing dependencies in evidence.
Is there any way, some logic in the construction of this project and its build that can be followed, anything to get me off the ground?

**Comment by LucasCelestinoSE** [comment]
<h1>The code that i want to build using cmake </h1>
<h2> II can't use the cmake you made, because it's a big build, I just need to use an example, I just need to build an example case and test using EMSCRIPTEN/WebAssembly.</h2>
`

                        #include <iostream>
                        #include <memory>
                        #include <vector>
                        #include "open_spiel/spiel.h"
                        #include "open_spiel/games/tic_tac_toe/tic_tac_toe.h"
                        
                        using namespace std;
                        using namespace open_spiel;

                      int main() {
                         auto game = LoadGame("tic_tac_toe");
                         auto state = game->NewInitialState();
                      
                          while (!state->IsTerminal()) {
                             cout << "Tabuleiro atual:\n" << state->ToString() << endl;
                      
                             Player current_player = state->CurrentPlayer();
                            cout << "Turno do jogador " << current_player << endl;
                      
                            vector<Action> legal_actions = state->LegalActions();
                            Action action = legal_actions[rand() % legal_actions.size()];
                      
                            cout << "Jogador " << current_player << " jogou na posição " << action << endl;
                            state->ApplyAction(action);
                      }
                      
                       return 0;
                      
                      }
`

**Reply by lanctot** [reply, marked as answer]
I assume you have built OpenSpiel as a shared library using the [instructions here](https://github.com/google-deepmind/open_spiel/blob/master/docs/library.md).

Then you can have a fairly simple `CMakeLists.txt` for the program above, i.e. it would look quite close to [this example](https://cliutils.gitlab.io/modern-cmake/chapters/basics/example.html).

It's probably best to go through the [CMake Tutorial](https://cmake.org/cmake/help/latest/guide/tutorial/index.html). There's also a nice [video tutorial here](https://www.youtube.com/watch?v=nlKcXPUJGwA&list=PLalVdRk2RC6o5GHu618ARWh0VO0bFlif4).

If, after learning CMake you have a problem compiling the basic program above, please follow-up here. For more complex questions regarding other technologies (like WebAssembly etc.) probably best to ask on their forums. For CMake in particular, there's a subreddit that may be useful too: https://www.reddit.com/r/cmake/
